### DISCUSSION TITLE
[Question] Is there a way to specify the output directory of shared libraries during compilation (instead of installing)

### DISCUSSION THREAD

**Discussion by Makogan** [mesonbuild/meson#10356] [Q&A]
I am aware that the `install_dir` argument allows the user to specify custom installation locations for shared libraries. This however requires one to do both meson compile and meson install when testing a project.

Moreover the intended final installation path for a project may be completely different from the path when testing. I.e. a normal workflow is compile -> run executable -> run tests -> edit -> compile.

When the above is done then one may wish to install.

I have read the documentation but I cannot find anything specifying if you can tell meson where to out the final shared binaries. By default meson duplicates the directory structure of the `meson.build` scripts and outputs all the shared binaries to the different directories that match where the original meson.build that specified the binary was.

However it is very common to release all shared objects alongside the executable, for example for plugin systems that allow runtime selection of features.

Is this possible at all?

**Comment by eli-schwartz** [comment]
In general no, you cannot use a kwarg to change the output directory for shared libraries any more than you can do so for other types of build outputs.

Meson does set up rpath entries to make sure that build targets can find their dependencies by default while uninstalled. The other traditional method would be setting LD_LIBRARY_PATH.


You may also want to check out `meson devenv` for setting up an interactive shell process with various environment variables set up to smooth the process using/running/building against the project while uninstalled. This is the recommended way to handle things rather than massaging the output directory.

(There is a function to define arbitrary variables that will be set in a devenv, to augment the ones that meson auto detects.)
